### ISSUE TITLE
Significant Performance Drop at the Lastest Version

### ISSUE THREAD

**Issue by vwxyzjn** [wandb/client#621]
`wandb --version && python --version && uname`

* Weights and Biases version: 0.8.13
* Python version: 3.7
* Operating System: Linux

### Description

After I installed the latest version of `wandb==0.8.13`, there is a significant performance drop compared to the previous version `wandb==0.8.1`. I was running it on a docker container `vwxyzjn/code-server:latest`, but in my local machine, such performance drop does not happen.

The performance drop is pretty drastic. In the weight and biases Web UI, I see the script finished maybe 400 timesteps in 30 secs, while after I reinstalled the older version `wandb==0.8.1`, The script finished more than 15,000 timesteps in 30 secs. Not sure why.

### What I Did

```python
# a2c_wandb_debug.py
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.distributions.categorical import Categorical
from torch.distributions.normal import Normal
from torch.utils.tensorboard import SummaryWriter

import numpy as np
from gym.spaces import Discrete, Box, MultiBinary, MultiDiscrete, Space

def preprocess_obs_space(obs_space: Space):
    if isinstance(obs_space, Discrete):
        return (obs_space.n,
                lambda x, obs_space=obs_space: F.one_hot(torch.LongTensor(x), obs_space.n).float())

    elif isinstance(obs_space, Box):
        return (np.array(obs_space.shape).prod(),
                lambda x, obs_space=obs_space: torch.Tensor(x).float().view(torch.Tensor(x).shape[0], -1))
def preprocess_ac_space(ac_space: Space, stochastic=True):
    if isinstance(ac_space, Discrete):
        return ac_space.n

    elif isinstance(ac_space, MultiDiscrete):
        return ac_space.nvec.sum()

    elif isinstance(ac_space, Box):
        # first half of the logits for mean and second half for std
        return np.prod(ac_space.shape) * 2

import argparse
import numpy as np
import gym
import pybullet_envs
from gym.spaces import Discrete, Box, MultiBinary, MultiDiscrete, Space
import time
import random

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='A2C agent')
    # Common arguments
    parser.add_argument('--exp-name', type=str, default="a2c",
                       help='the name of this experiment')
    parser.add_argument('--gym-id', type=str, default="BipedalWalker-v2",
                       help='the id of the gym environment')
    parser.add_argument('--learning-rate', type=float, default=7e-4,
                       help='the learning rate of the optimizer')
    parser.add_argument('--seed', type=int, default=1571109047,
                       help='seed of the experiment')
    parser.add_argument('--episode-length', type=int, default=200,
                       help='the maximum length of each episode')
    parser.add_argument('--total-timesteps', type=int, default=4000000,
                       help='total timesteps of the experiments')
    parser.add_argument('--torch-deterministic', type=bool, default=True,
                       help='whether to set `torch.backends.cudnn.deterministic=True`')
    parser.add_argument('--prod-mode', type=bool, default=False,
                       help='run the script in production mode and use wandb to log outputs')
    parser.add_argument('--wandb-project-name', type=str, default="cleanRL",
                       help="the wandb's project name")
    
    # Algorithm specific arguments
    parser.add_argument('--gamma', type=float, default=0.99,
                       help='the discount factor gamma')
    parser.add_argument('--vf-coef', type=float, default=0.25,
                       help="value function's coefficient the loss function")
    parser.add_argument('--max-grad-norm', type=float, default=0.5,
                       help='the maximum norm for the gradient clipping')
    parser.add_argument('--ent-coef', type=float, default=0.01,
                       help="policy entropy's coefficient the loss function")
    args = parser.parse_args()
    if not args.seed:
        args.seed = int(time.time())

# TRY NOT TO MODIFY: setup the environment
env = gym.make(args.gym_id)
random.seed(args.seed)
np.random.seed(args.seed)
torch.manual_seed(args.seed)
torch.backends.cudnn.deterministic = args.torch_deterministic
env.seed(args.seed)
env.action_space.seed(args.seed)
env.observation_space.seed(args.seed)
input_shape, preprocess_obs_fn = preprocess_obs_space(env.observation_space)
output_shape = preprocess_ac_space(env.action_space)

# ALGO LOGIC: initialize agent here:
class Policy(nn.Module):
    def __init__(self):
        super(Policy, self).__init__()
        self.fc1 = nn.Linear(input_shape, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, output_shape)

    def forward(self, x):
        x = preprocess_obs_fn(x)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

class Value(nn.Module):
    def __init__(self):
        super(Value, self).__init__()
        self.fc1 = nn.Linear(input_shape, 64)
        self.fc2 = nn.Linear(64, 1)

    def forward(self, x):
        x = preprocess_obs_fn(x)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return x

pg = Policy()
vf = Value()
optimizer = optim.Adam(list(pg.parameters()) + list(vf.parameters()), lr=args.learning_rate)
loss_fn = nn.MSELoss()

# TRY NOT TO MODIFY: start the game
experiment_name = f"{args.gym_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
writer = SummaryWriter(f"runs/{experiment_name}")
writer.add_text('hyperparameters', "|param|value|\n|-|-|\n%s" % (
        '\n'.join([f"|{key}|{value}|" for key, value in vars(args).items()])))
if args.prod_mode:
    import wandb
    wandb.init(project=args.wandb_project_name, tensorboard=True, config=vars(args), name=experiment_name)
    writer = SummaryWriter(f"/tmp/{experiment_name}")
global_step = 0
while global_step < args.total_timesteps:
    next_obs = np.array(env.reset())
    actions = np.empty((args.episode_length,), dtype=object)
    rewards, dones = np.zeros((2, args.episode_length))
    obs = np.empty((args.episode_length,) + env.observation_space.shape)
    
    # ALGO LOGIC: put other storage logic here
    values = torch.zeros((args.episode_length))
    neglogprobs = torch.zeros((args.episode_length,))
    entropys = torch.zeros((args.episode_length,))
    
    # TRY NOT TO MODIFY: prepare the execution of the game.
    for step in range(args.episode_length):
        global_step += 1
        obs[step] = next_obs.copy()
        
        # ALGO LOGIC: put action logic here
        logits = pg.forward([obs[step]])
        values[step] = vf.forward([obs[step]])

        # ALGO LOGIC: `env.action_space` specific logic
        if isinstance(env.action_space, Discrete):
            probs = Categorical(logits=logits)
            action = probs.sample()
            actions[step], neglogprobs[step], entropys[step] = action.tolist()[0], -probs.log_prob(action), probs.entropy()

        elif isinstance(env.action_space, Box):
            mean, std = torch.split(logits, int(output_shape/2), dim=1)
            probs = Normal(mean, torch.abs(std))
            action = probs.sample()
            action = torch.clamp(action, torch.min(torch.Tensor(env.action_space.low)), torch.min(torch.Tensor(env.action_space.high)))
            actions[step], neglogprobs[step], entropys[step] = action.tolist()[0], -probs.log_prob(action).sum(), probs.entropy().sum()
    
        elif isinstance(env.action_space, MultiDiscrete):
            logits_categories = torch.split(logits, env.action_space.nvec.tolist(), dim=1)
            action = []
            probs_categories = []
            probs_entropies = torch.zeros((logits.shape[0]))
            neglogprob = torch.zeros((logits.shape[0]))
            for i in range(len(logits_categories)):
                probs_categories.append(Categorical(logits=logits_categories[i]))
                if len(action) != env.action_space.shape:
                    action.append(probs_categories[i].sample())
                neglogprob -= probs_categories[i].log_prob(action[i])
                probs_entropies += probs_categories[i].entropy()
            action = torch.stack(action).transpose(0, 1).tolist()
            actions[step], neglogprobs[step], entropys[step] = action[0], neglogprob, probs_entropies
        
        # TRY NOT TO MODIFY: execute the game and log data.
        next_obs, rewards[step], dones[step], _ = env.step(actions[step])
        next_obs = np.array(next_obs)
        if dones[step]:
            break
    
    # ALGO LOGIC: training.
    # calculate the discounted rewards, or namely, returns
    returns = np.zeros_like(rewards)
    for t in reversed(range(rewards.shape[0]-1)):
        returns[t] = rewards[t] + args.gamma * returns[t+1] * (1-dones[t])
    # advantages are returns - baseline, value estimates in our case
    advantages = returns - values.detach().numpy()
    
    vf_loss = loss_fn(torch.Tensor(returns), torch.Tensor(values)) * args.vf_coef
    pg_loss = torch.Tensor(advantages) * neglogprobs
    loss = (pg_loss - entropys * args.ent_coef).mean() + vf_loss
    
    optimizer.zero_grad()
    loss.backward()
    nn.utils.clip_grad_norm_(list(pg.parameters()) + list(vf.parameters()), args.max_grad_norm)
    optimizer.step()

    # TRY NOT TO MODIFY: record rewards for plotting purposes
    writer.add_scalar("charts/episode_reward", rewards.sum(), global_step)
    writer.add_scalar("losses/value_loss", vf_loss.item(), global_step)
    writer.add_scalar("losses/entropy", entropys.mean().item(), global_step)
    writer.add_scalar("losses/policy_loss", pg_loss.mean().item(), global_step)
env.close()
writer.close()
```

```bash
$ pip install wandb==0.8.1
$ python a2c_wandb_debug.py \
    --seed 1 \
    --gym-id CartPole-v0 \
    --total-timesteps 30000 \
    --wandb-project-name cleanrl.benchmark \
    --prod-mode True
$ pip install wandb --upgrade
$ python a2c_wandb_debug.py \
    --seed 1 \
    --gym-id CartPole-v0 \
    --total-timesteps 30000 \
    --wandb-project-name cleanrl.benchmark \
    --prod-mode True
```

**Comment by raubitsj** [comment]
@vwxyzjn, Thanks for the code sample!

I tried to reproduce but I have not been successful.  Here are my two runs:
0.8.1: https://app.wandb.ai/jeffr/cleanrl.benchmark/runs/93zrfwxx/overview
0.8.13: https://app.wandb.ai/jeffr/cleanrl.benchmark/runs/sl4s2tk6/overview

They both finished all 30000 steps in 33 and 39 seconds respectively

Can you post your wandb run URLs so I can look closer into what might be causing this issue in your environment.

**Comment by vanpelt** [comment]
If you can send us the Dockerfile that would help.  My gut says there’s an
unoptimized json or yaml library installed in the environment as that’s the
only meaningful operation we do in the user script.

**Comment by vwxyzjn** [comment]
Hi,

Upon further investigation, there actually wasn't a performance drop, but for some reasons the charts don't display all the y-axis, thus creating an illusion of performance drop. See

`# pip install wandb==0.8.1`
https://app.wandb.ai/costa-huang/cleanrl.benchmark/runs/jpq1pdx5?workspace=user-costa-huang
`# pip install wandb --upgrade`
https://app.wandb.ai/costa-huang/cleanrl.benchmark/runs/nerlx4cf?workspace=user-costa-huang

As you can see, the 0.8.1 displays all the 1k step, while 0.8.13 displays 43 step.

 the container is 
[vwxyzjn/code-server:2.1600-vsc1.38.1](https://hub.docker.com/layers/vwxyzjn/code-server/2.1600-vsc1.38.1/images/sha256-ae583b1a0cd87b8f6b7728616085c97a7a802a47a0b5511c71ee26e8d1031418)
And the Dockerfile is as follow 

```Dockerfile
# docker build -t vwxyzjn/code-server:2.1600-vsc1.38.1 -t vwxyzjn/code-server:latest .

FROM codercom/code-server:2.1600-vsc1.38.1

USER root

# Install Go Dep
RUN wget -q https://github.com/golang/dep/releases/download/v0.4.1/dep-linux-amd64 -O /usr/bin/dep
RUN chmod +x /usr/bin/dep
# Install Go
RUN apt-get update
RUN apt-get install -y wget git gcc
RUN wget --quiet -P /tmp https://dl.google.com/go/go1.12.4.linux-amd64.tar.gz
RUN tar -C /usr/local -xzf /tmp/go1.12.4.linux-amd64.tar.gz
RUN rm /tmp/go1.12.4.linux-amd64.tar.gz
ENV GOPATH /home/coder/project/go
ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"

# Install anaconda
ENV LANG=C.UTF-8 LC_ALL=C.UTF-8
ENV PATH /opt/conda/bin:$PATH
RUN apt-get update --fix-missing && apt-get install -y wget bzip2 ca-certificates \
    libglib2.0-0 libxext6 libsm6 libxrender1 \
    git mercurial subversion
RUN wget --quiet https://repo.anaconda.com/archive/Anaconda3-2019.10-Linux-x86_64.sh -O ~/anaconda.sh && \
    /bin/bash ~/anaconda.sh -b -p /opt/conda && \
    rm ~/anaconda.sh && \
    ln -s /opt/conda/etc/profile.d/conda.sh /etc/profile.d/conda.sh && \
    echo ". /opt/conda/etc/profile.d/conda.sh" >> ~/.bashrc && \
    echo "conda activate base" >> ~/.bashrc

# Install python tools
RUN pip install dacite
RUN sudo apt install -y mpich default-jdk htop
RUN conda install pytorch torchvision cpuonly -c pytorch
RUN pip install gym[atari,box2d]
RUN pip install stable_baselines
RUN pip install tensorflow==1.15.0
RUN pip install git+https://github.com/vwxyzjn/baselines
RUN pip install pybullet
```
The docker-compose.yaml that runs the env is 
```docker-compose.yaml
version: '2'

services:
  code-server:
    image: vwxyzjn/code-server:latest
    volumes:
      - data-volume:/home/coder/project
      - /var/run/docker.sock:/var/run/docker.sock
    restart: always
    environment:
      - PASSWORD=${PASSWORD}
    command: --auth password --allow-http --disable-telemetry /home/coder/project
    networks:
      - web
    labels:
      - "traefik.docker.network=web"
      - "traefik.enable=true"
      - "traefik.basic.frontend.rule=Host:dcode.costa.sh"
      - "traefik.basic.port=8080"
      - "traefik.basic.protocol=http"

networks:
  web:
    external: true

volumes:
  data-volume:
```

**Comment by vanpelt** [comment]
I see what's going, we changed the way we log tensorboard events in 0.8.10.  Before we were storing the global_step in the default x-axis.  In the new release we store it in a custom x-axis named "global_step".  If you click the "X" icon next to the search bar you can change the x-axis to "global_step" and both charts will look the same.  We're discussing internally if this is something we want to do by default to remove this confusion.

**Comment by vwxyzjn** [comment]
I see. That fixes it. Perhaps remove one or the other, otherwise, two steps axis are kind of confusing. Thanks for the support.
